### PR TITLE
When releasing futures except AttributeError for early closing clients

### PR DESCRIPTION
**Pull request by quasiben**
PR resolves https://github.com/dask/distributed/issues/4305

I think this is a bit better than checking for `AttributeError`.  If you have time, @martindurant I'd appreciate your thoughts here

### CODE REVIEW

**Comment by martindurant** [comment]
This seems eminently sensible to me.
Can it be triggered in a test, assign a Future to a client, then close the scheduler before calling `release()` explicitly? I've has a 2-minute attempt at this and didn't find a way to make it fail. If you can't do it in 5min, then we should merge this.

**Comment by quasiben** [comment]
I haven't been able to trigger on my laptop.  Just on the DGX :/ I'll take a look some more tomorrow

File: distributed/client.py [hunk 1]
```diff
@@ -353,7 +353,11 @@ def type(self):
...
         # (see e.g. Client.get() or Future.__del__())
-        if not self._cleared and self.client.generation == self._generation:
+        if (
+            not self._cleared
+            and self.client.generation == self._generation
+            and self.client.scheduler is not None
+        ):
             self._cleared = True
             try:
                 self.client.loop.add_callback(self.client._dec_ref, stringify(self.key))
```
Comment (mrocklin) [inline distributed/client.py:363]: The _dec_ref method does other things, like reducing the current refcount on the client.  If we're closing then sure, I'm happy for us not to this, but if we're in a state where we might connect again then this seems not quite right.

I'm curious, how does this situation come about?

If we're woried about the scheduler connection being down, then maybe there is a more comprehensive / central place to handle this, like in `_send_to_scheduler`.

Comment (martindurant) [reply distributed/client.py:363]: I believe the alternative is to catch multiple exceptions in `Future.release()`. 
The ignored exception is seen during interpreter shutdown, when it seems the client has closed its comm first. I don't think its been seen in ordinary cleanup of futures.

Comment (quasiben) [reply distributed/client.py:363]: I originally saw this issue in streamz test suite but only when running on DGXs.  I can reproduce on my laptop.  With @martindurant help, I was able to build a [reproducer](https://github.com/dask/distributed/issues/4305) which only relied on dask/distributed but only occurs when exiting the program or closing all the connections

**Comment by martindurant** [comment]
@quasiben , do you want to simplify by adding the exception to the `except` clause in `Future.__del__`, instead of this solution, with a comment that we think this only happens during interpreter shutdown.

**Comment by quasiben** [comment]
> @quasiben , do you want to simplify by adding the exception to the except clause in Future.__del__, instead of this solution, with a comment that we think this only happens during interpreter shutdown.

Sure, I'll do that

**Comment by quasiben** [comment]
This keeps failing in various Travis tests.  I am restarting CI again

**Comment by martindurant** [comment]
when do we squash Travis?! :)

**Comment by quasiben** [comment]
Update:

I have restarted this job a number of times -- it keeps failing on:

> distributed/tests/test_worker.py::test_wait_for_outgoing

Locally, this passes.  Restarting CI again

**Comment by quasiben** [comment]
Grr, this time it failed on:

> test_broken_worker_during_computation

And another one segfaulted.  Restarting again

**Comment by mrocklin** [comment]
That one is known to fail on Travis

**Comment by mrocklin** [comment]
I think that there is also a check somewhere of if we're shutting down the Python process.  I think that you might want `distributed.utils.shutting_down`

File: distributed/client.py [hunk 2]
```diff
@@ -382,10 +382,11 @@ def __setstate__(self, state):
     def __del__(self):
         try:
             self.release()
-        except AttributeError:
+        except AttributeError as e:
             # Ocassionally we see this error when shutting down the client
             # https://github.com/dask/distributed/issues/4305
-            pass
+            if not shutting_down():
+                raise e
```
Comment (mrocklin) [inline distributed/client.py]: I think you want to drop the `as e` here and just call `raise`

If done within an except block this raises the original exception with a clean traceback I think.

Comment (quasiben) [reply distributed/client.py]: thanks @mrocklin -- fixed in #0455fa8

**Pull request merged**